---
title: ""
author: "Guo-Bo Chen"
date: "`r Sys.Date()`"
output:
 html_document:
    theme: united
    highlight: tango
    code_folding: hide
params:
  froot: arab
  espace: 2
  sc: 1
  pcut: 0.05
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
#source("helper.R")
```

```{r, manhattan}
manhattan <- function(dataframe, colors=c("gray10", "gray50"), ymax="max", limitchromosomes=NULL, suggestiveline=-log10(1e-5), genomewideline=NULL, title="", annotate=NULL, ...) {
  
  d=dataframe
  if (!("CHR" %in% names(d) & "BP" %in% names(d) & "P" %in% names(d))) stop("Make sure your data frame contains columns CHR, BP, and P")
  if (!is.null(limitchromosomes)) {
    d=d[d$CHR %in% limitchromosomes, ]
  }
  
  d=subset(na.omit(d[order(d$CHR, d$BP), ]), (P>0 & P<=1)) # remove na's, sort, and keep only 0<P<=1
  d$logp = -log10(d$P)
  d$pos=NA
  ticks=NULL
  lastbase=0 #  colors <- rep(colors,max(d$CHR))[1:max(d$CHR)]
  colors <- rep(colors,max(d$CHR))[1:length(unique(d$CHR))]
  
  if (ymax=="max") ymax<-ceiling(max(d$logp))
  if (ymax<8) ymax<-8
  numchroms=length(unique(d$CHR))
  if (numchroms==1) {
    d$pos=d$BP
    ticks=floor(length(d$pos))/2+1
  } else {
    Uchr=unique(d$CHR)
    for (i in 1:length(Uchr)) {
      if (i==1) {
        d[d$CHR==Uchr[i], ]$pos=d[d$CHR==Uchr[i], ]$BP
      } else {
        lastbase=lastbase+tail(subset(d, CHR==Uchr[i-1])$BP, 1)
        d[d$CHR==Uchr[i], ]$pos=d[d$CHR==Uchr[i], ]$BP+lastbase
      }
      ticks=c(ticks, d[d$CHR==Uchr[i], ]$pos[floor(length(d[d$CHR==Uchr[i], ]$pos)/2)+1])
    }
  }
  if (numchroms==1) {
    with(d, plot(main=title, pos, logp, ylim=c(0,ymax), ylab=expression(-log[10](italic(p))), xlab=paste("Chromosome",unique(d$CHR),"position"), ...))
  } else {
    with(d, plot(main=title, pos, logp, ylim=c(0,ymax), ylab=expression(-log[10](italic(p))), xlab="Chromosome", xaxt="n", type="n", ...))
    axis(1, at=ticks, lab=unique(d$CHR), ...)
    icol=1
    Uchr=unique(d$CHR)
    for (i in 1:length(Uchr)) {
      with(d[d$CHR==Uchr[i], ], points(pos, logp, col=colors[icol], ...))
      icol=icol+1
    }
  }
  if (!is.null(annotate)) {
    d.annotate=d[which(d$SNP %in% annotate), ]
    with(d.annotate, points(pos, logp, col="green3", ...))
  }
  #  if (suggestiveline) abline(h=suggestiveline, col="blue")
  if (!is.null(genomewideline)) {
    abline(h=genomewideline, col="gray")
  } else {
    abline(h=-log10(0.05/nrow(d)), col="gray")    
  }
}
```


# EigenGWAS Report {.tabset .tabset-fade .tabset-pills}
A brief summary for EigenGWAS analysis.

## 1 MAF
Minor allele frequency distribution of the sample.
```{r, maf}
fq=read.table(paste0(params$froot, ".frq"), as.is = T, header = T)
hist(fq$MAF, main="MAF", xlab="Minor allele frequency", xlim=c(0, 0.5), breaks = 50)
```

## 2 PCA projection
The distribution of the top eigenvalues and the projection of the samples in the first two eigenspace.
```{r, pca}
layout(matrix(1:2, 1, 2))
evalF=read.table(paste0(params$froot, ".eigenval"), as.is = T)
barplot(evalF[,1]/params$sc, border = F, main="Eigenvalue")
abline(h=1, lty=2, col="black")

pcF=read.table(paste0(params$froot, ".eigenvec"), as.is = T)
plot(main="eSpace 1 vs 2", pcF[,3], pcF[,4], xlab="eSpace 1", ylab="eSpace 2", bty='n', pch=16, cex=0.5, col=ifelse(pcF[,3]<0, "red", "blue"))
```

## 3 GRM
$n_e$ reflects the relative sample size of the sample. If the sample is random, the $n_e$ should be very close to the sample size of the data. $m_e$ is the effective number of markers for the sample, reflecting the average linkage disequilibrium between markers.
```{r, grm}
layout(matrix(1:2, 1, 2))
gz=gzfile(paste0(params$froot, ".grm.gz"))
grm=read.table(gz, as.is = T)
Ne=-1/mean(grm[grm[,1]!=grm[,2], 4]/params$sc)
Me=1/var(grm[grm[,1]!=grm[,2], 4]/params$sc)
hist(grm[grm[,1]!=grm[,2],4]/params$sc, main="Pairwise relatedness ", xlab="Relatedness score", breaks = 50)
        
nn=nrow(read.table(paste0(params$froot, ".fam"), as.is = T))
mm=nrow(read.table(paste0(params$froot, ".bim"), as.is = T))
legend("topright", legend = c(paste0("ne=", format(Ne, digits=3, nsmall=2), ' [',nn, ']'), paste0("me=", format(Me, digits=3, nsmall=2), ' [',mm,']')), bty='n')

hist(grm[grm[,1]==grm[,2],4]/params$sc, main="Diagonal relatedness", xlab="Relatedness score", breaks = 15)
```

## 4 Eigenvalue vs $\lambda_{GC}$
Eigenvalue reflects the mean of selection, and $\lambda_{GC}$ reflects the median of the selection intensity.
```{r, ev}
PC=params$espace
Evev=read.table(paste0(params$froot, ".eigenval"), as.is = T)
GC=array(0, dim=PC)
for(i in 1:PC) {
  eg = read.table(paste0(params$froot, ".", i, ".assoc.linear"), as.is = T, header = T)
  GC[i] = qchisq(median(eg$P), 1, lower.tail = F)/qchisq(0.5, 1)
}

egc=matrix(c(Evev[1:PC,1]/params$sc, GC), PC, 2, byrow = F)
rownames(egc)=seq(1, PC)
barplot(t(egc), beside = T, border = F, xlab="eSpace", ylim=c(0,max(egc)+2))
        abline(h=1, lty=2, lwd=2)
legend("topright", legend = c("Eigenvalue", expression(paste(lambda[gc]))), pch=15, col=c("black", "grey"), bty='n')

```

## 5 EigenGWAS
Manhhattan plot for genomic scan for eigenvectors.
```{r, eGWAS}
PC=params$espace
froot = params$froot

for(i in 1:PC) {
  layout(matrix(1:2, 1, 2))
  EigenRes=read.table(paste0(froot, ".", i, ".assoc.linear"), as.is = T, header = T)
  EigenRes$Praw=EigenRes$P
  gc=qchisq(median(EigenRes$P), 1, lower.tail = F)/qchisq(0.5, 1, lower.tail = F)
  EigenRes$P=pchisq(qchisq(EigenRes$Praw, 1, lower.tail = F)/gc, 1, lower.tail = F)
  manhattan(EigenRes, genomewideline = -log10(params$pcut/nrow(EigenRes)), title=paste("eSpace ", i), pch=16, cex=0.3, bty='n')

  #QQplot
  chiseq=qchisq(seq(1/nrow(EigenRes), 1-1/nrow(EigenRes), length.out = nrow(EigenRes)), 1)
  qqplot(chiseq, qchisq(EigenRes$Praw, 1, lower.tail = F), xlab=expression(paste("Theoretical ", chi[1]^2)), ylab=expression(paste("Observed ", chi[1]^2)), bty="n", col="grey", pch=16, cex=0.5)
  points(sort(chiseq), sort(qchisq(EigenRes$P, 1, lower.tail = F)), col="black", pch=16, cex=0.5)
  legend("topleft", legend = c("Raw", "GC correction"), pch=16, cex=0.5, col=c("grey", "black"), bty='n')
  abline(a=0, b=1, col="red", lty=2)
}

```
